wpiutil/src/main/native/include/wpi/jni_util.h

```diff
@@ -175,207 +175,237 @@ class JStringRef {
+    m_size = other.m_size;
+    m_elements = other.m_elements;
+    other.m_valid = false;
+    other.m_jarr = nullptr;
+    other.m_elements = nullptr;
+    return *this;
+  }
 
-  std::string_view str() const {
-    auto arr = static_cast<const C*>(this)->array();
-    if (arr.empty()) {
-      return {};
+  JSpanBase(JNIEnv* env, jobject bb, size_t size)
+    requires(!IsCritical)
+      : m_valid{Size == std::dynamic_extent || size == Size},
+        m_env{env},
+        m_jarr{nullptr},
+        m_size{size},
+        m_elements{static_cast<std::remove_cv_t<T>*>(
+            bb ? env->GetDirectBufferAddress(bb) : nullptr)} {
+    if (!bb) {
+      errs() << "JSpan was passed a null pointer at \n"
+             << GetJavaStackTrace(env);
     }
-    return {reinterpret_cast<const char*>(arr.data()), arr.size()};
   }
 
-  std::span<const uint8_t> uarray() const {
-    auto arr = static_cast<const C*>(this)->array();
-    if (arr.empty()) {
-      return {};
+  JSpanBase(JNIEnv* env, jarray_type jarr, size_t size)
+      : m_valid{Size == std::dynamic_extent || size == Size},
+        m_env{env},
+        m_jarr{jarr},
+        m_size{size},
+        m_elements{nullptr} {
+    if (jarr) {
+      if constexpr (IsCritical) {
+        m_elements = static_cast<std::remove_cv_t<T>*>(
+            env->GetPrimitiveArrayCritical(jarr, nullptr));
+      } else {
+        m_elements = ArrHelper::GetArrayElements(env, jarr);
+      }
+    } else {
+      errs() << "JSpan was passed a null pointer at \n"
+             << GetJavaStackTrace(env);
     }
-    return {reinterpret_cast<const uint8_t*>(arr.data()), arr.size()};
   }
-};
 
-/**
- * Specialization of JArrayRefBase to handle both "long long" and "long" on
- * 64-bit systems.
- */
-template <typename C>
-class JArrayRefInner<C, jlong> {
- public:
-  template <typename U>
-    requires(sizeof(U) == sizeof(jlong) && std::integral<U>)
-  operator std::span<const U>() const {  // NOLINT
-    auto arr = static_cast<const C*>(this)->array();
-    if (arr.empty()) {
-      return {};
+  JSpanBase(JNIEnv* env, jarray_type jarr)
+      : JSpanBase(env, jarr, jarr ? env->GetArrayLength(jarr) : 0) {}
+
+  ~JSpanBase() {
+    if (m_jarr && m_elements) {
+      constexpr jint mode = std::is_const_v<T> ? JNI_ABORT : 0;
+      if constexpr (IsCritical) {
+        m_env->ReleasePrimitiveArrayCritical(m_jarr, m_elements, mode);
+      } else {
+        ArrHelper::ReleaseArrayElements(m_env, m_jarr, m_elements, mode);
+      }
     }
-    return {reinterpret_cast<const U*>(arr.data()), arr.size()};
   }
-};
 
-/**
- * Base class for J*ArrayRef and CriticalJ*ArrayRef
- */
-template <typename T>
-class JArrayRefBase : public JArrayRefInner<JArrayRefBase<T>, T> {
- public:
-  explicit operator bool() const { return this->m_elements != nullptr; }
+  operator std::span<T>() const { return array(); }
 
-  std::span<const T> array() const {
-    if (!this->m_elements) {
+  std::span<T> array() const {
```

Comment:
```suggestion
  std::span<T, Size> array() const {
```

And ditto other cases?  I'm not sure what the purpose of the Size parameter is otherwise.  Although this won't work with the error return case, so we'd have to find a different solution there.  Should we just remove the Size template parameter entirely?

Reply:
Yeah, I tried to specify the size of the span but ran into the issue you mentioned that we can't use the default constructor for sized spans.
We could maybe remove the Size template parameter, but it does provide a nice way to check the runtime size of the array against a specified constant size, so it does have some purpose (though we could change the check to being between the size parameter to the `(JNIEnv*, jarray_type, size_t)` constructor and the `env->GetArrayLength` of the array, which would have the side effect of adding the check to the few direct uses of that constructor (ntcore generated types_jni and wpiutil DataLogJNI)).

Reply:
std::span has a functional default constructor for sized spans. The pointer is nullptr and the size of zero.

Reply:
We could put the `!m_elements` check in an `if constexpr (Size != dynamic_extent)` block. The return would be UB if m_elements was nullptr, but that's probably ok?

Reply:
It looks like `m_elements` will only be null if constructed from a null pointer (in which case it writes to stderr with the JavaStackTrace) or it's moved from. If it's constructed from a null pointer, it writes to stderr with the Java stacktrace, so UB there is \*probably\* okay, and I don't think we do any moving, so UB there is \*probably\* also okay. (Plus most of the spans are dynamic extent anyways)